float32 large_strength
float32 small_strength